File: src/main/java/models/FoodItem.java
package models;

import java.util.List;

/**
 * class stores food item details
 */
public class FoodItem {
    private int id;
    private String name;
    private double carbohydrates;
    private double fat;
    private double protein;
    private double calories;
    private List<ServingSize> servingSizes;

    /**
     * constructor
     *
     * @param id
     * @param name
     * @param carbohydrates
     * @param fat
     * @param protein
     * @param servingSizes
     */
    public FoodItem(int id, String name, double carbohydrates, double fat, double protein, List<ServingSize> servingSizes) {
        this.id = id;
        this.name = name;
        this.carbohydrates = carbohydrates;
        this.fat = fat;
        this.protein = protein;
        this.calories = 4 * carbohydrates + 9 * fat + 4 * protein;
        this.servingSizes = servingSizes;
    }

    /**
     * getter for fdcId
     *
     * @return String
     */
    public int getId() {
        return id;
    }

    /**
     * getter for food description
     *
     * @return String
     */
    public String getName() {
        return name;
    }

    /**
     * getter for carbohydrates
     *
     * @return double
     */
    public double getCarbohydrates() {
        return carbohydrates;
    }

    /**
     * getter for fat
     *
     * @return double
     */
    public double getFat() {
        return fat;
    }

    /**
     * getter for protein
     *
     * @return double
     */
    public double getProtein() {
        return protein;
    }

    /**
     * getter for calories
     *
     * @return double
     */
    public double getCalories() {
        return Math.round(calories * 100.0) / 100.0;
    }

    /**
     * getter for serving size
     *
     * @return List<ServingSize>
     */
    public List<ServingSize> getServingSizes() {
        return servingSizes;
    }

    /**
     * override default toString method
     *
     * @return String
     */
    @Override
    public String toString() {
        return "FoodItem{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", carbohydrates=" + carbohydrates +
                ", fat=" + fat +
                ", protein=" + protein +
                ", calories=" + calories +
                ", servingSizes=" + servingSizes +
                '}';
    }

}

File: client/src/Components/FoodItem.js
import React, {Component} from 'react'
import update from "immutability-helper"
import {SERVER_URL} from "../config";
import ServingSize from "./ServingSize";
import ItemDeleteButton from "./ItemDeleteButton";

class FoodItem extends Component {
    constructor(props) {
        super(props);
        this.state = {
            adding: false,
            deleting: false,
            loading: false,
            item: this.props.completedItem ? this.props.completedItem : undefined
        }
    }

    handleServingSizeChange(newServingSizeId) {
        let newServingSize = this.state.item.servingSizes.find(servingSize => servingSize.id === newServingSizeId);
        let newState = update(this.state, {
            item: {
                selectedServing: {
                    servingSize: {$set: newServingSize}
                }
            }
        });
        this.setState(newState);
    }

    handleQuantityChange(newServingQuantity) {
        let newState = update(this.state, {
            item: {
                selectedServing: {
                    quantity: {$set: newServingQuantity}
                }
            }
        });
        this.setState(newState);
    }

    handleItemClick() {
        if (!this.props.editMode) {
            if (!this.state.adding && !this.state.item) {
                this.setState({loading: true})
                fetch(`${SERVER_URL}` + "/api/foods/" + this.props.item.id)
                    .then((response) => response.json)
                    .then(item => {
                        let processedItem = this.setDefaultServingQuantity(item);
                        this.setState({
                            adding: true,
                            item: processedItem,
                            loading: false
                        })
                    });
            } else {
                this.setState(prevState => ({
                    adding: !prevState.adding
                }));
            }
        } else {
            this.setState(prevState => ({
                deleting: !prevState.deleting
            }));
        }
    }

    /**
     * set initial serving quantity to 100 gram for foods
     * whose only serving size is 1 gram
     * @param item
     * @returns {*}
     */
    setDefaultServingQuantity(item) {
        if (item.selectedServing.servingSize.label === 'g' && item.selectedServing.quantity === 1) {
            item.selectedServing.quantity = 100;
        }
        return item;
    }

    calculateItemTotals() {
        console.log("hi")
        console.log(this.state.item)
        let item = this.state.item;
        let totalCalories, totalCarbohydrates, totalFat, totalProtein;
        totalCalories = totalCarbohydrates = totalFat = totalProtein = 0;
        let selectedServing = item.servingSizes[0];
        let servingSizeMultiplier = 1 * selectedServing.ratio;

        totalCalories += Math.round(item.calories * servingSizeMultiplier);
        totalCarbohydrates += Math.round(item.carbohydrates * servingSizeMultiplier);
        totalFat += Math.round(item.fat * servingSizeMultiplier);
        totalProtein += Math.round(item.protein * servingSizeMultiplier);

        return {
            calories: totalCalories,
            carbohydrates: totalCarbohydrates,
            fat: totalFat,
            protein: totalProtein
        };
    }

    deleteUserFoodItem() {
        this.props.deleteUserFoodItem(this.state.item.id);
    }

    render() {
        let foodName;
        if (this.props.completedItem) {
            foodName = this.state.item.name;
        } else if (this.props.item) {
            foodName = this.props.item.name;
        } else {
            foodName = 'Name'
        }

        let itemTotals;
        if (this.state.item) {
            itemTotals = this.calculateItemTotals();
        }

        return (<div
            className={'FoodItem' + (this.state.adding ? ' adding' : '') + (this.state.deleting ? ' deleting' : '')}
            onClick={this.handleItemClick.bind(this)}>
            <span>{foodName}</span>
            {this.state.item &&
            <span>
                <span>
                    <span>{itemTotals.carbohydrates}</span>
                    <span>{itemTotals.fat}</span>
                    <span>{itemTotals.protein}</span>
                </span>
                <span>{itemTotals.calories}</span>
            </span>
            }

            {this.state.adding && this.state.item &&
            <ServingSize
                selectedServing={this.state.item.selectedServing}
                servingSizes={this.state.item.servingSizes}
                itemId={this.state.item.id}
                handleQuantityChange={this.handleQuantityChange.bind(this)}
                handleSizeChange={this.handleServingSizeChange.bind(this)}
            />}

            {this.state.deleting && <ItemDeleteButton deleteItem={this.deleteUserFoodItem.bind(this)}/>}
        </div>);
    }
}

export default FoodItem;